imap_processing/lo/l1b/lo_l1b.py

```diff
@@ -260,6 +266,127 @@ def set_spin_cycle(l1a_de: xr.Dataset, l1b_de: xr.Dataset) -> xr.Dataset:
     return l1b_de
 
 
+def get_spin_start_times(
+    l1a_de: xr.Dataset, l1b_de: xr.Dataset, spin_data: xr.Dataset, acq_end: xr.DataArray
+) -> xr.DataArray:
+    """
+    Get the start time for the spin that each direct event is in.
+
+    The resulting array of spin start times will be equal to the length of the direct
+    events. If two direct events occurred in the same spin, then there will be repeating
+    spin start times.
+
+    Parameters
+    ----------
+    l1a_de : xr.Dataset
+        The L1A DE dataset.
+    l1b_de : xr.Dataset
+        The L1B DE dataset.
+    spin_data : xr.Dataset
+        The L1A Spin dataset.
+    acq_end : xr.DataArray
+        The end acquisition times for each spin ASC.
+
+    Returns
+    -------
+    spin_start_time : xr.DataArray
+        The start time for the spin that each direct event is in.
+    """
+    met = l1a_de["met"].values
+    # Find the closest stop_acq for each shcoarse
+    closest_stop_acq_indices = np.abs(met[:, None] - acq_end.values).argmin(axis=1)
+    # There are 28 spins per epoch (1 aggregated science cycle)
+    # Set the spin_cycle_num to the spin number relative to the
+    # start of the ASC
+    spin_cycle_num = l1b_de["spin_cycle"] % 28
+    # Get the seconds portion of the start time for each spin
+    start_sec_spins = spin_data["start_sec_spin"].values[
+        closest_stop_acq_indices, spin_cycle_num
+    ]
```

Comment:
Can you stay within xarray and subset based on your dimensions something like this? I find this a bit more explicit what dimensions you're using and subsetting on.
```suggestion
    spin_subset = spin_data.isel(epoch=closest_stop_acq_indices, spin_number=spin_cycle_num)
    spin_start_time = spin_subset["start_sec_spin"] + spin_subset["start_subsec_spin"] * 1e-6
```

Reply:
I might not be understanding how to use `isel`, but it looks like that returns a 2D array filtered using the epoch and spin indices used in the function. The indexing I'm using returns a 1D array selecting a single spin from the epoch for that DE.

Reply:
I see, I missed that. And now I just read this excellent tutorial explaining the differences between numpy and xarray indexing defaults and how to do it with xarray if you do want that behavior from numpy by default too by creating a new dimension to index along: https://tutorial.xarray.dev/intermediate/indexing/advanced-indexing.html

What you have is good here 👍